typescript/looker/rtl/authSession.ts

```diff
@@ -0,0 +1,113 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import { IAuthSession, IRequestProps, ITransport } from './transport'
+import { IApiSettings } from './apiSettings'
+
+export abstract class AuthSession implements IAuthSession {
+  static TBD = "Method not implemented in AuthSession"
+  settings: IApiSettings
+  sudoId: string = ''
+  transport: ITransport
+
+  constructor(settings: IApiSettings, transport: ITransport) {
+    this.settings = settings
+    this.transport = transport
+  }
+
+  /**
+   * Decorate request properties with the required authentication information
+   *
+   * Properties could be additional headers, cookies, or some other request property
+   *
+   * @param props Request properties to use and/or modify in authentication method
+   */
+  abstract authenticate(props: IRequestProps): Promise<IRequestProps>
+
+  /**
+   * Does the session have active authentication information?
+   */
+  abstract isAuthenticated() : boolean
+
+  /**
+   * Override this method to implement API session login
+   *
+   * @param sudoId ID of sudo user. A missing sudoId means the API credentials are used to login
+   */
+  // @ts-ignore sudoId is not used in this default implementation
+  login(sudoId?: string | number): Promise<any> {
+    return this.notImplementedAsync()
+  }
+
+  /**
+   * Override this method to implement logout
+   *
+   * The base implementation doesn't throw a not implemented error. It just returns a false promise.
+   */
+  logout(): Promise<boolean> {
+    return new Promise<boolean>(() => false)
+  }
+
+  /**
+   * Override this method to implement a authentication retrieval from the server
+   *
+   * The base implementation will throw an error in the promise result saying the method isn't implemented
+   *
+   */
+  getToken(): Promise<any> {
+    return this.notImplementedAsync()
+  }
+
+  /**
+   * Is a sudo user active for this session?
+   */
+  isSudo(): boolean {
+    return (this.sudoId !== '') && (this.isAuthenticated())
+  }
+
+  /**
+   * Resets all authentication status, but standard implementations do NOT log out an API session
+   */
+  reset(): void {
+    this.sudoId = ''
+  }
+
+  /**
+   * utility function not implemented error
+   */
+  private notImplemented() {
+    throw new Error(AuthSession.TBD)
+  }
+
+  /**
+   * utility function not implemented error promise
+   */
+  private notImplementedAsync() {
+    // Use `ctrl` for instance closure in the method reference
+    const ctrl = this
```

Comment:
`ctrl` just provides a captured reference for `this` because the `this` reference can get ambiguous inside a lambda. At least, I've run into that issue before. Maybe @lukelooker can confirm whether or not the `ctrl` closure is needed in this instance.

Reply:
Ok, so I spent more time than I'd like to admit trying to understand this better. I'm not sure just assigning a new local variable to use as a closure in the constructor arg for the `Promise` would accomplish the binding that you're looking for. If `notImplementedAsync()` ends up being called on a different object then `ctrl/this` still reference that other object: [I think I demonstrated that here](http://www.typescriptlang.org/play/?ssl=23&ssc=63&pln=1&pc=1#code/MYGwhgzhAEBiD29oG8BQ0PQA4CcCWAbmAC4Cm0AdvMQJIC2WIpdpFZAJgBQCUK6mA4gAsc8AO6VSEgKI5ROTgHJEWCIu78MAX03YArgCMQeYNEgBPCsABy1eo2asOPPgIHB4FCMWjBiOEGgAXmhhPAhdARxSYj0cCkkJAAVROnDSAB4wCnMAPk4XINzffxAAOipaBiYWNlIubg0BHR1UPDqcADMwYHIAeWFSHFdMAHN4dtGAFXgp-GAAawBNeD0ALmhC4pT4NIhM7LzdSvsap3qNregCCfZUVo8vH07gxLhEHlRUR+9oaiEhhsBgDhiE0AJxpMZnMTMtVhtOmULFZbFUHLUOAAaY52aqOOrsBF6KzEPCeFzIUIicRvWTyJT+WH1aDmVbqaAtL7-IZlSEUaazeZwvQ8MrAEjAIQFaKQTy8Iq+TwQeBMMogeCjTgy5UURpAA)

However, if `notImplementedAsync()` is called on the instance of the `AuthSession` subclass then `ctrl/this` is captured by the arrow function. So I believe the `ctrl/this` assignment doesn't do anything and if you want to protect against using the wrong `this` you'd maybe have to do something with `bind`

All this said, I have very little knowledge/exposure to the subtleties of the javascript `this` so I could be totally of base here but I'd love to learn and better understand it.

Reply:
A key property of arrow functions (which are used here) is that they bind this in an expected way, so I believe you do not need this rebinding (and in fact I think ESLint would disallow it). TS should also through a type error if you were calling a method on the wrong object.

Reply:
Thanks, Wil. I'll take that closure out, and I'm also going to write some unit tests before committing more changes to the PR, so I will verify the behavior as well.